### PR TITLE
ReentrantReadWriteLock replace ReentrantLock in LRUCache

### PR DESCRIPTION
**Pull request by dugenkui03**
## What is the purpose of the change

#2610 

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java [hunk 1]
```diff
@@ -41,75 +45,90 @@ public LRUCache(int maxCapacity) {
...
             super.clear();
         } finally {
-            lock.unlock();
+            writeLock.unlock();
         }
     }
 
     public int getMaxCapacity() {
-        return maxCapacity;
+        readLock.lock();
```
[the remark this answers is not in the document]
Comment (dugenkui03) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: > why should add lock here?

Just to keep the same style：original code add lock on method `size()`-this method just read variable in HashMap(`transient int size`).this method just read variable `maxCapacity`.

I think it's safe to remove lock,and that can improve the performance.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java [hunk 2]
```diff
@@ -27,7 +27,11 @@
     private static final float DEFAULT_LOAD_FACTOR = 0.75f;
 
     private static final int DEFAULT_MAX_CAPACITY = 1000;
-    private final Lock lock = new ReentrantLock();
+
+    private final ReentrantReadWriteLock rwlock = new ReentrantReadWriteLock(true);
```
[the remark this answers is not in the document]
Comment (dugenkui03) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: > Code style.
> Use RW_LOCK.
not `static final`,and i think `lock` can't be static.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (carryxyh) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: agree.
unnecessary lock. :)

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (carryxyh) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: agree with that can not be static.
but use upper with final field is better.
what do u think?

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dugenkui03) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: > agree.
> unnecessary lock. :)

>i ignored that  maxCapacity is volatile,and it is  threadSafe without lock.and size. size() with lock can assure memory visibility [参考](https://stackoverflow.com/questions/12429818/does-explicit-lock-automatically-provide-memory-visibility).i will fix it.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dugenkui03) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: > agree with that can not be static.
> but use upper with final field is better.
> what do u think?

This is also OK.Code style referred to the jdk before.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (carryxyh) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: > i ignored that maxCapacity is volatile,

Yep, that is why I have questions with lock with this method.

**Comment by carryxyh** [comment]
LGTM.
@beiwei30 
Would u pls review this pr?

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dugenkui03) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java]: > > i ignored that maxCapacity is volatile,
> 
> Yep, that is why I have questions with lock with this method.

Thx for your help. Code has been updated

**Comment by dugenkui03** [comment]
> LGTM.
> @beiwei30
> Would u pls review this pr?

`get()` method in `LinkedHashMap`  is not threadsafe.that's a bad pr.
[details](https://blog.csdn.net/wawmg/article/details/19482041)

复盘最近PR为明天面试做准备的时候，突然发现这里不能用RWLock，因为其数据结构基础`linkedMap`的`get()`方法将会修改保持顺序的链表，即`get()`操作也有修改。感谢花费时间帮助review。

**Comment by carryxyh** [comment]
My fault.
Thx for your contribution, too.

**Pull request closed without merging**